prepare-local-docker-env.sh

```diff
@@ -1,47 +1,71 @@
-  cargo build-bpf --bpf-out-dir ./here
-  mv ./here/spl_associated_token_account.so $CWD/programs/ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL.so
-popd
-
-pushd solana_program_library/token/program
-  cargo build-bpf --bpf-out-dir ./here
-  mv ./here/spl_token.so $CWD/programs/TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA.so
-popd
-
-pushd solana_program_library/token/program-2022
-  cargo build-bpf --bpf-out-dir ./here
-  mv ./here/spl_token_2022.so $CWD/programs/TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb.so
-popd
-
-rm -rf solana_program_library
-rm -rf metaplex_program_library
+#!/bin/bash
+
+# output colours
+RED() { echo $'\e[1;31m'$1$'\e[0m'; }
+GRN() { echo $'\e[1;32m'$1$'\e[0m'; }
+
+CURRENT_DIR=$(pwd)
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+# go to parent folder
+cd $(dirname $(dirname $SCRIPT_DIR))
+
+EXTERNAL_ID=("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s" \
+"BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY" \
+"cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK" \
+"noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV" \
+"ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL" \
+"TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA" \
+"TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb" \
+)
+EXTERNAL_SO=("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s.so" \
+"BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY.so" \
+"cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK.so" \
+"noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV.so"
+"ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL.so" \
+"TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA.so" \
+"TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb.so" \
+)
+
+if [ -z ${RPC+x} ]; then
+    RPC="https://api.mainnet-beta.solana.com"
+fi
+
+if [ -z "$OUTPUT" ]; then
+    OUTPUT=$CURRENT_DIR/programs
+fi
+
+# creates the output directory if it doesn't exist
+if [ ! -d ${OUTPUT} ]; then
+    mkdir ${OUTPUT}
+fi
+
+# only prints this if we have external programs
+if [ ${#EXTERNAL_ID[@]} -gt 0 ]; then
+    echo "Dumping external programs to: '${OUTPUT}'"
+fi
+
+# dump external programs binaries if needed
+for i in ${!EXTERNAL_ID[@]}; do
+    if [ ! -f "${OUTPUT}/${EXTERNAL_SO[$i]}" ]; then
+        solana program dump -u $RPC ${EXTERNAL_ID[$i]} ${OUTPUT}/${EXTERNAL_SO[$i]}
+    else
+        solana program dump -u $RPC ${EXTERNAL_ID[$i]} ${OUTPUT}/onchain-${EXTERNAL_SO[$i]} > /dev/null
+        ON_CHAIN=`sha256sum -b ${OUTPUT}/onchain-${EXTERNAL_SO[$i]} | cut -d ' ' -f 1`
+        LOCAL=`sha256sum -b ${OUTPUT}/${EXTERNAL_SO[$i]} | cut -d ' ' -f 1`
+
+        if [ "$ON_CHAIN" != "$LOCAL" ]; then
+            echo $(RED "[ WARNING ] on-chain and local binaries are different for '${EXTERNAL_SO[$i]}'")
+        else
+            echo "$(GRN "[ SKIPPED ]") on-chain and local binaries are the same for '${EXTERNAL_SO[$i]}'"
+        fi
+
+        rm ${OUTPUT}/onchain-${EXTERNAL_SO[$i]}
+    fi
+done
+
+# only prints this if we have external programs
+if [ ${#EXTERNAL_ID[@]} -gt 0 ]; then
+    echo ""
```

Comment:
Shouldn't we have some message here? Like `External programs dumped to ${OUTPUT}`

Reply:
Yes, it does print some messages like that.  So this `echo ""` is only a formatting to give one whitespace if you printed a list.  The output when tehre's external programs looks like this with one blank newline:
```
danenbm@danenbm-thelio:~/Metaplex-Workspace/digital-asset-rpc-infrastructure$ ./prepare-local-docker-env.sh 
Dumping external programs to: '/home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs'
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA.so
Wrote program to /home/danenbm/Metaplex-Workspace/digital-asset-rpc-infrastructure/programs/TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb.so

danenbm@danenbm-thelio:~/Metaplex-Workspace/digital-asset-rpc-infrastructure$ 
```

Reply:
I think its just the stdout of the `solana program dump` that does it

Reply:
I don't think you should change it, it was mostly out of curiosity, about why that line was there.